design/ceph/osd-migration.md

```diff
@@ -0,0 +1,230 @@
+- `ceph-volume` commands use the `OSD_STORE_FLAG` while activating the OSDs.
+
+    - RAW MODE
+      ```
+      ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs
+      ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"
+      ```
+
+**NOTE**: ceph-volume `activate` command doesn't require `OSD_STORE_FLAG` to be passed as an argument. It auto detects the backend store that was used during the OSD `prepare`
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a Ceph version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  store:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### API
+
+```go
+// OSDReplaceInfo represents an OSD that needs to replaced
+type OSDReplaceInfo struct {
+	ID   int    `json:"id"`
+	Path string `json:"path"`
+	Node string `json:"node"`
+}
+
+// ReplaceQueue contains the list of OSDs to be replaced one at a time
+type ReplaceQueue struct {
+	OSDs []OSDReplaceInfo
+}
+
+// OSDReplaceConfig contains the status of the OSD replacement
+type OSDReplaceConfig struct {
+	OSDs                *ReplaceQueue    `json:"osds"`
+	IsPgHealthy          bool            `json:"isPgHealthy"`
+	OSDInProgress        *OSDReplaceInfo `json:"OSDInProgress"`
+	StartTimeStamp       time.Time       `json:"startTimeStamp"`
+}
+```
+
+**OSDReplaceInfo**: Represents an OSD that needs to be replaced
+  - **ID**: OSD ID of the OSD to be replaced.
+  - **Path**: Disk or pvc claim that the OSD is consuming.
+  - **Node**: Node on which the OSD is running
+
+
+**ReplaceQueue**: This is a queue implementation containing a list of OSDs to be replaced. It will support basic queue operations like push, pop, size etc.
+
+
+**OSDReplaceConfig**: This will hold the current status/progress of the OSD replacement in the cluster.
+  - **OSDs**:  List of OSDs that are yet to be replaced.
+  - **IsPgHealth**: Whether Placement Groups are healthy or not. This is a boolean field.
+  - **OSDInProgress**: Represents the OSDs for which the deployment is being deleted.
+  - **StartTimeStamp**: Time at which the whole OSD replacement process was started. This will help users know how much time has elapsed since the first OSD was replaced.
+
+
+#### Replace OSD
+
+Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label not matching the osd store type provided in `spec.storage.store.type`. This list will be stored in `OSDReplaceConfig.OSDs`
```

Comment:
Actually, do we even need to store the list of OSD IDs in a configmap? Every reconcile we can just query the OSD labels and then choose one of them to convert. Maybe the only thing we need to store in a configmap is the OSD in progress. Everything else is just internal implementation for the operator or osd prepare job.

Logging the list of OSD IDs in need of conversion would be helpful though in case the admin wants to see progress in detail besides the cluster CR status.

Reply:
What if we add the list of OSDs in need of migration to the CephCluster status? Then we can keep the configmaps as an internal implementation detail and not try to use them as an API.

Reply:
I gave it another thought.  The list of pending OSDs might not be very useful information. A simple flag in the `status.ceph` in the `cephCluster` yaml will be more useful instead. 

For example: 
``` status:
    ceph:
      capacity:
        bytesAvailable: 32147042304
        bytesTotal: 32212254720
        bytesUsed: 65212416
        lastUpdated: "2023-08-07T08:01:36Z"
      details:
        OSD_SLOW_PING_TIME_BACK:
          message: Slow OSD heartbeats on back (longest 2080.270ms)
          severity: HEALTH_WARN
        OSD_SLOW_PING_TIME_FRONT:
          message: Slow OSD heartbeats on front (longest 2090.234ms)
          severity: HEALTH_WARN
      fsid: d6cbe2a9-1648-47fd-a5e8-5889596aa344
      health: HEALTH_WARN
      lastChanged: "2023-08-07T08:01:36Z"
      lastChecked: "2023-08-07T08:01:36Z"
      previousHealth: HEALTH_OK
      versions:
        mgr:
          ceph version 17.2.6-373-g5406f4b1 (5406f4b1729bf4f476338ad344cc86b126e0c5b3) quincy (stable): 1
        mon:
          ceph version 17.2.6-373-g5406f4b1 (5406f4b1729bf4f476338ad344cc86b126e0c5b3) quincy (stable): 3
        osd:
          ceph version 17.2.6-373-g5406f4b1 (5406f4b1729bf4f476338ad344cc86b126e0c5b3) quincy (stable): 3
        overall:
          ceph version 17.2.6-373-g5406f4b1 (5406f4b1729bf4f476338ad344cc86b126e0c5b3) quincy (stable): 7
     osdStoreMigration: <COMPLETED/INPROGRESS>
```

What do you think?